### PR TITLE
MESOS: fix panic in controller-manager when apiserver lookups fail

### PR DESCRIPTION
**Pull request by s-urbaniak**
xref https://github.com/mesosphere/kubernetes-mesos/issues/656

### CODE REVIEW

File: contrib/mesos/pkg/node/statusupdater.go [hunk 1]
```diff
@@ -61,46 +61,62 @@ func (u *StatusUpdater) Run(terminate <-chan struct{}) error {
...
+	go runtime.Until(monitor, u.heartBeatPeriod, terminate)
 	return nil
 }
 
+// filter is a helper alias around a slice of node objects.
+type filter []api.Node
+
+// having returns a slice of node objects matching
+// the given slaves as their Spec.ExternalIDs.
+func (f filter) having(slaves []string) []api.Node {
```
Comment (sttts) [inline contrib/mesos/pkg/node/statusupdater.go]: whichAreIn ? having doesn't make sense for two sets.

File: contrib/mesos/pkg/node/statusupdater.go [hunk 2]
```diff
@@ -61,46 +61,62 @@ func (u *StatusUpdater) Run(terminate <-chan struct{}) error {
...
-			err := u.updateStatus(&nodes[i])
+		nodes := nl.(*api.NodeList).Items
+
+		// for nodes *having* a kubelet:
+		// let the kubelet do its job updating the status, or the
+		// node controller will remove this node if the node does not even
+		// exist anymore
+
+		for _, n := range filter(nodes).having(slavesWithoutKubelet) {
+			// update status for nodes which do not have a kubelet running and
```
Comment (sttts) [inline contrib/mesos/pkg/node/statusupdater.go]: Would move that above the loop.

File: contrib/mesos/pkg/node/statusupdater.go [hunk 3]
```diff
@@ -61,46 +61,62 @@ func (u *StatusUpdater) Run(terminate <-chan struct{}) error {
...
-				// let the kubelet do its job updating the status, or the
-				// node controller will remove this node if the node does not even
-				// exist anymore
-				continue
-			}
-
-			err := u.updateStatus(&nodes[i])
+		nodes := nl.(*api.NodeList).Items
+
+		// for nodes *having* a kubelet:
```
Comment (sttts) [inline contrib/mesos/pkg/node/statusupdater.go]: Would combine this comment with the lower one.

File: contrib/mesos/pkg/node/statusupdater.go (hunk 1 above)
Comment (s-urbaniak) [inline contrib/mesos/pkg/node/statusupdater.go]: as discussed we chose `intersect`

**Comment by s-urbaniak** [comment]
PTAL @sttts @jdef

**Comment by s-urbaniak** [comment]
While fixing the bug me and @sttts just recognized that querying via the listwatch is inefficient in the current implementation:
https://github.com/kubernetes/kubernetes/blob/master/contrib/mesos/pkg/node/statusupdater.go#L78

The suggestion is to use `nodeStore` instead. This makes the unit tests of the set-jazz pretty unnecessary.

The following commit implements the suggestion mentioned here.

**Comment by jdef** [comment]
:saxophone:

**Comment by s-urbaniak** [comment]
@jdef PTAL
Additionally i smoke-tested this patch locally using:
```
$ NUM_NODES=3 ./cluster/kube-up.sh
$ kubectl describe no
```
On two machines a kubelet has been started marked as `KubeletReady`, one machine was being left idle and correctly marked as `SlaveReady` which is what the code in this commit does.

**Comment by jdef** [comment]
lgtm. please squash

**Comment by s-urbaniak** [comment]
@jdef squashed :-)

**Pull request merged**